src/domain/balances/balances.repository.ts

```diff
@@ -83,12 +84,23 @@ export class BalancesRepository implements IBalancesRepository {
     simpleBalances.map((simpleBalance) =>
       this.simpleBalancesValidator.validate(simpleBalance),
     );
+
+    const tokenAddresses = simpleBalances
+      .map((simpleBalance) => simpleBalance.tokenAddress)
+      .filter((address): address is string => address !== null);
+
+    const prices = tokenAddresses.length
+      ? await this._getTokenPrices(args.chainId, args.fiatCode, tokenAddresses)
+      : [];
+
     return await Promise.all(
       simpleBalances.map(async (simpleBalance) => {
+        const tokenAddress = simpleBalance.tokenAddress?.toLowerCase() ?? null;
         const price = await this._getPrice(
           args.chainId,
           args.fiatCode,
-          simpleBalance.tokenAddress,
+          tokenAddress,
+          prices,
```

Comment:
Maybe this is more related with naming and overall structure but I find it a bit confusing that we get Token Prices (`_getTokenPrices`), for all the tokens in the balances, and then we call `_getPrice` for each token in the balances "again". I know that the functionality is different but pointing out that semantically it was a bit confusing.

Reply:
Yes, I think that's a valid point. The first `_getTokenPrices` get the prices for all the tokens in advance, and then the second `_getPrice` _selects_ a price from that list. I agree with you this is a bit confusing semantically, I think we could just rename the second function. But I'm leaning towards just removing the private `_getPrice` function, as it's pretty simple, it has few lines, and its purpose is a bit blurry, so we can integrate it into the caller body.

Reply:
I will leave this up to you. I'm fine either way as long as the differences between the two "get prices" are understandable.

Reply:
I've integrated the function logic in its caller in [5a27454](https://github.com/safe-global/safe-client-gateway/pull/836/commits/5a274541184ed56b79ed8f5cffefc259092cc027)